src/expr/impl/src/scalar/concat.rs

```diff
@@ -0,0 +1,73 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::fmt::Write;
+
+use risingwave_common::row::Row;
+use risingwave_common::types::ToText;
+use risingwave_expr::function;
+
+/// Concatenates the text representations of all the arguments. NULL arguments are ignored.
+///
+/// # Example
+///
+/// ```slt
+/// query T
+/// select concat('abcde', 2, NULL, 22);
+/// ----
+/// abcde222
+///
+/// query T
+/// select concat(variadic array['abcde', '2', NULL, '22']);
+/// ----
+/// abcde222
+/// ```
+#[function("concat(variadic anyarray) -> varchar")]
+fn concat(vals: impl Row, writer: &mut impl Write) {
```

Comment:
I'm wondering what the possible receiver types are here for `impl Row`. 👀

Reply:
It is `RowRef` for non-variadic version, and `ListRef` for variadic version.